File: lincoln/ruby/lib/types/table/basic_table.rb
require "lib/types/table/table"
require "lib/types/basic/tuple_set"
require "lib/types/table/hash_index"

class BasicTable < Table
  def initialize(context, name, key, types)
    @tuples = TupleSet.new(name, nil)
    super(name, TableType::TABLE, key, types)
    @key = key
    @primary = HashIndex.new(context, self, key, Index::Type::PRIMARY)
    @secondary = Hash.new
    @context = context
  end
  
  def tuples
    return (@tuples.nil? ? TupleSet.new(name) : @tuples.clone)
  end
  
  attr_reader :primary, :secondary

  def cardinality
    @tuples.size
  end
  
  def insert_tup(tuple)
    @tuples << tuple
    return tuples
  end
  def delete_tup(tuple)
    return @tuples.delete(tuple)
  end
end

File: lincoln/java/src/jol/types/table/BasicTable.java
package jol.types.table;

import java.util.Hashtable;
import jol.types.basic.Tuple;
import jol.types.basic.TupleSet;
import jol.types.basic.TypeList;
import jol.types.exception.UpdateException;
import jol.core.Runtime;

public class BasicTable extends Table {
	/* The primary key. */
	protected Key key;
	
	protected TupleSet tuples;
	
	protected Index primary;
	
	protected Hashtable<Key, Index> secondary;
	
	public BasicTable(Runtime context, TableName name, Key key, TypeList types) {
		super(name, Type.TABLE, key, types);
		this.key = key;
		this.tuples = new TupleSet(name);
		this.primary = new HashIndex(context, this, key, Index.Type.PRIMARY);
		this.secondary = new Hashtable<Key, Index>();
	}
	
	@Override
	public TupleSet tuples() {
		try {
		return this.tuples == null ? new TupleSet(name()) : this.tuples.clone();
		} catch (Exception e) {
			System.err.println("TABLE " + name() + " ERROR: " + e);
			e.printStackTrace();
			System.exit(0);
		}
		return null;
	}
	
	@Override
	protected boolean insert(Tuple t) throws UpdateException {
		return this.tuples.add(t);
	}
	
	@Override
	protected boolean delete(Tuple t) throws UpdateException {
		return this.tuples.remove(t);
	}

	@Override
	public Index primary() {
		return this.primary;
	}

	@Override
	public Hashtable<Key, Index> secondary() {
		return this.secondary;
	}

	@Override
	public Integer cardinality() {
		return this.tuples().size();
	}

}
